[Location tracker] - Fix some location tracking issues (PSF-1000)

mnaturel opened the pull request:
<!-- Please read [CONTRIBUTING.md](https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md) before submitting your pull request -->
 
## Type of change

- [ ] Feature
- [ ] Bugfix
- [ ] Technical
- [x] Other : Enhancement

## Content

<!-- Describe shortly what has been changed -->
- Use the latest known location among the ones from all providers when starting tracking.
- Adding minimal period of update to notify callback of location change to avoid too many notifications.
- Prioritize locations received from the providers in the following order: “Fused”, “GPS”, other providers
- Avoid calling callback for provider not available when only one of them is disabled (should be called only when all providers are disabled)


## Motivation and context

<!-- Provide link to the corresponding issue if applicable or explain the context -->
Closes #5913 

## Screenshots / GIFs

<!-- Only if UI have been changed
You can use a table like this to show screenshots comparison.
Uncomment this markdown table below and edit the last line `|||`:
|copy screenshot of before here|copy screenshot of after here|
-->

<!--
|Before|After|
|-|-|
|||
 -->

## Tests

<!-- Explain how you tested your development -->

- Check location sharing is enabled in the settings
- Go to a room
- Start to share location
- Check the current location is good on the map
- Check the logs of the LocationTracker

## Tested devices

- [x] Physical
- [x] Emulator
- OS version(s): Android 11

## Checklist

<!-- Depending on the Pull Request content, it can be acceptable if some of the following checkboxes stay unchecked. -->

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

bmarty commented on vector/src/main/java/im/vector/app/features/location/LocationTracker.kt:
```diff
@@ -127,6 +128,19 @@ class LocationTracker @Inject constructor(
+        val elapsedMillis = nowMillis - (lastLocationUpdateMillis ?: 0)
+        if (elapsedMillis >= MIN_TIME_TO_UPDATE_LOCATION_MILLIS) {
+            lastLocationUpdateMillis = nowMillis
+            notifyLocation(location)
```
Nit You could use our handy `FirstThrottler`.

mnaturel replied:
Thanks for the info. In fact, I realized a debouce operation is more suitable here. I will try to rework on it.

bmarty commented on vector/src/main/java/im/vector/app/features/location/LocationTracker.kt:
```diff
@@ -55,34 +57,36 @@ class LocationTracker @Inject constructor(
+
+                        locationManager.getLastKnownLocation(provider)
+                    }
+                    .maxByOrNull { location -> location.time }
```
👍

bmarty reviewed (commented):
Just a small remark

ganfra commented on vector/src/main/java/im/vector/app/features/location/LocationTracker.kt:
```diff
@@ -36,62 +42,90 @@ class LocationTracker @Inject constructor(
         if (locationManager == null) {
-            callbacks.forEach { it.onLocationProviderIsNotAvailable() }
-            Timber.v("## LocationTracker. LocationManager is not available")
+            callbacks.forEach { it.onNoLocationProviderAvailable() }
```
Maybe try catch the dispatching

ganfra commented on vector/src/main/java/im/vector/app/features/location/LocationTracker.kt:
```diff
@@ -36,62 +42,90 @@ class LocationTracker @Inject constructor(
-                    // Send last known location without waiting location updates
-                    locationManager.getLastKnownLocation(provider)?.let { lastKnownLocation ->
+        if (providers.isEmpty()) {
+            callbacks.forEach { it.onNoLocationProviderAvailable() }
```
same

mnaturel replied:
Done, see [e21b696](https://github.com/vector-im/element-android/pull/5923/commits/e21b6963c0ea53ff4a4760dcf06dc798049b3939)

ganfra commented on vector/src/main/java/im/vector/app/features/location/LocationTracker.kt:
```diff
@@ -36,62 +42,90 @@ class LocationTracker @Inject constructor(
+            Timber.v("There is no location provider available")
+        } else {
+            // Take GPS first
+            providers.sortedByDescending { getProviderPriority(it) }
```
use method reference?

mnaturel replied:
Done, see [af0ae61](https://github.com/vector-im/element-android/pull/5923/commits/af0ae6123b9a7a9a1249cd9a1c672db9268fef0f). Should we prioritize this syntax as a general rule?

ganfra replied:
We might yes!

ganfra commented on vector/src/main/java/im/vector/app/features/location/LocationTracker.kt:
```diff
@@ -110,41 +144,67 @@ class LocationTracker @Inject constructor(
         val locationData = location.toLocationData()
         lastLocation = locationData
-        callbacks.forEach { it.onLocationUpdate(locationData) }
+        callbacks.forEach { it.onLocationUpdate(location.toLocationData()) }
```
Same, try/catch?

mnaturel replied:
Done, see [e21b696](https://github.com/vector-im/element-android/pull/5923/commits/e21b6963c0ea53ff4a4760dcf06dc798049b3939)

ganfra reviewed (commented):
Just small remarks otherwise LGTM and thanks for the tests!

mnaturel commented:
I have fixed a regression I introduced concerning the stopping of the location tracking: [0e11e4b](https://github.com/vector-im/element-android/pull/5923/commits/0e11e4b2a6d25c02b21b8ec35adacc506657cec4)

ganfra reviewed (approved):
LGTM!

Claire1817 commented on vector/src/test/java/im/vector/app/features/location/LocationTrackerTest.kt:
```diff
@@ -0,0 +1,302 @@
+    }
+
+    @After
+    fun tearDown() {
```
Maybe you should unmockk the objects after each tests ? 
Location or LocationTracker for exemple ?

mnaturel replied:
> Maybe you should unmockk the objects after each tests ? Location or LocationTracker for exemple ?

I am not sure to correctly understand your suggestion. Do you suggest to create a new `LocationTracker` object for each test? I didn't think about it but if so, I agree so that we can keep each test isolated from each other.

mnaturel replied:
I have done an update here: [abe7fe9](https://github.com/vector-im/element-android/pull/5923/commits/abe7fe917615b588a1e932ce6a30373cda9312c3)

bmarty commented on vector/src/main/java/im/vector/app/features/location/LocationTracker.kt:
```diff
@@ -117,34 +151,84 @@ class LocationTracker @Inject constructor(
+    }
+
+    private fun onNoLocationProviderAvailable() {
+        callbacks.forEach {
```
Is there a risk of `ConcurrentModificationException` here?

mnaturel replied:
Hum I guess there is... I will try to add a test to see if it passes.

mnaturel replied:
Updated here: [1ee852d](https://github.com/vector-im/element-android/pull/5923/commits/1ee852dbf4e223ebab5a7de775ba4e2c14de3b6c)

bmarty commented on vector/src/main/java/im/vector/app/features/location/LocationTracker.kt:
```diff
@@ -117,34 +151,84 @@ class LocationTracker @Inject constructor(
+    }
+
+    private fun onLocationUpdate(locationData: LocationData) {
+        callbacks.forEach {
```
Same remark.

bmarty reviewed (commented):
Just one latest remark.

bmarty reviewed (approved):
(Did not perform a full review, but my remarks have been handled)
